### PR TITLE
feat(internal): use autoReplace for docker managers

### PR DESCRIPTION
**Pull request by rarkins**

### CODE REVIEW

File: lib/datasource/docker/index.ts [hunk 1]
```diff
@@ -43,6 +43,8 @@ export const defaultConfig = {
   group: {
     commitMessageTopic: '{{{groupName}}} Docker tags',
   },
+  autoReplaceNewString:
+    '{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}',
```
Comment (rarkins) [inline lib/datasource/docker/index.ts:47]: Doing it like this is necessary to handle the case of digest pinning.

File: lib/manager/kubernetes/index.ts [hunk 2]
```diff
@@ -1,10 +1,11 @@
 import { LANGUAGE_DOCKER } from '../../constants/languages';
 
 export { extractPackageFile } from './extract';
-export { updateDependency } from './update';
 
 export const language = LANGUAGE_DOCKER;
 
+export const autoReplace = true;
```
Comment (viceice) [inline lib/manager/kubernetes/index.ts:7]: Flip default value and assume false if update function exists?

**Review by viceice — DISMISSED** [review]

File: lib/manager/kubernetes/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/manager/kubernetes/index.ts:7]: Yes, I'm going to do that eventually but will leave it as-is for now.

File: lib/workers/repository/process/fetch.ts [hunk 3]
```diff
@@ -30,6 +31,8 @@ async function fetchDepUpdates(
   const { depName, currentValue } = dep;
   // TODO: fix types
   let depConfig = mergeChildConfig(packageFileConfig, dep);
+  const datasourceDefaultConfig = await getDefaultConfig(depConfig.datasource);
+  Object.assign(dep, datasourceDefaultConfig);
```
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/process/fetch.ts:35]: IIRC the problem is that we need to edit the `dep` object in-place

File: lib/workers/branch/auto-replace.ts [hunk 4]
```diff
@@ -88,6 +85,7 @@ export async function doAutoReplace(
     currentDigest,
     newDigest,
     autoReplaceData,
+    autoReplaceNewString,
```
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/branch/auto-replace.ts:88]: Maybe. The original is `autoReplaceString` so I was attempting to make this name related. Maybe `autoReplaceStringTemplate` ?

[the remark this answers is not in the document]
Comment (viceice) [reply lib/workers/branch/auto-replace.ts:88]: Sure

**Review by viceice — APPROVED** [review]

File: lib/workers/repository/process/fetch.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/process/fetch.ts:35]: Thinking if this causes any problems. I think I need to do away with the `no-param-reassign` disabling

File: lib/workers/branch/auto-replace.spec.ts [hunk 5]
```diff
@@ -107,5 +107,26 @@ describe('workers/branch/auto-replace', () => {
...
+      upgrade.newValue = '8.11.4-alpine';
+      upgrade.currentDigest =
+        'sha256:d743b4141b02fcfb8beb68f92b4cd164f60ee457bf2d053f36785bf86de16b0d';
+      upgrade.newDigest = 'sha256:aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa';
+      upgrade.autoReplaceData = {
+        depIndex: 0,
+        replaceString:
+          'node:8.11.3-alpine@sha256:d743b4141b02fcfb8beb68f92b4cd164f60ee457bf2d053f36785bf86de16b0d',
+      };
+      upgrade.autoReplaceNewString =
```
Comment (viceice) [inline lib/workers/branch/auto-replace.spec.ts:126]:
```suggestion
      upgrade.autoReplaceStringTemplate =
```

**Review by viceice — CHANGES_REQUESTED** [review]

**Review by viceice — DISMISSED** [review]

File: lib/workers/repository/process/fetch.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (rarkins) [reply lib/workers/repository/process/fetch.ts]: Now changed

**Review by viceice — DISMISSED** [review]

**Pull request merged**